evm/src/generation/mpt.rs

```diff
@@ -33,13 +33,44 @@ impl Default for AccountRlp {
     }
 }
 
-#[derive(RlpEncodable, RlpDecodable, Debug)]
+#[derive(Clone, Debug)]
+pub enum AccessListInner {
+    List(Vec<AccessListItemRlp>),
+    Item(AccessListItemRlp),
+}
+
+impl Encodable for AccessListInner {
+    fn rlp_append(&self, s: &mut RlpStream) {
+        match self {
+            AccessListInner::List(list) => s.append_list(list),
+            AccessListInner::Item(item) => s.append(item),
+        };
+    }
+}
+
+impl Decodable for AccessListInner {
+    fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+        if rlp.is_list() {
```

Comment:
If this is in the hot loop it can probably be optimized a bit by checking `is_empty()` once and then switch based on byte 0 instead, like:

```rust
if rlp.is_empty() { return DecoderError:RlpExpectedToBeData }
// Can only be list or data now and byte 0 is != 0xc0
if rlp.as_raw[0] > 0xc0 { 
  // do list stuff 
} else {
  // do data stuff
}
```